### PR TITLE
Add @SideEffectFree to transfer function files

### PR DESCRIPTION
**Pull request by AndrewShf**
For methods in `CFAbstractTransfer.java`, `BackwardTransferFunction.java`, `ForwardTransferFunction.java` with no side effect, add SideEffectFree annotations.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 1]
```diff
@@ -490,6 +494,7 @@ private void addFinalLocalValues(S store, Element enclosingElement) {
      * @param methodDeclTree the declaration of the method or constructor
      * @return true if the receiver of a method or constructor might not yet be fully initialized
      */
+    @SideEffectFree
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:497]: Could this be Pure?

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 2]
```diff
@@ -147,6 +148,7 @@ protected CFAbstractTransfer(
      * @return true if the transfer function uses sequential semantics, false if it uses concurrent
      *     semantics
      */
+    @SideEffectFree
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:151]: Should this be Pure?

**Review by wmdietl — COMMENTED** [review]
Thanks!

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java (hunk 1 above)
Comment (AndrewShf) [reply framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:497]: SideEffectFree, Deterministic, Pure annotation are trusted annotations.

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 3]
```diff
@@ -162,6 +165,7 @@ public boolean usesSequentialSemantics() {
      * @param store the store
      * @return the possibly-modified value
      */
+    @Pure
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:168]: Shouldn't this, and the other `finishX` method below, stay only SideEffectFree? To allow a subclass to return a new object?

Comment (AndrewShf) [reply framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:168]: Okay, I forgot to consider about overriding.

**Review by wmdietl — APPROVED** [review]

**Pull request merged**